### PR TITLE
Move TLS CBC code into the providers

### PR DESCRIPTION
**Pull request by mattcaswell**
libssl has always managed its own CBC padding. Consequently there is constant time CBC padding removal code in libssl. This is also closely related to constant time code to remove the MAC from a MAC-then-encrypt record.

This PR moves the CBC padding and MAC removal code into the providers. This is important for the FIPS provider so that this code is moved inside the FIPS boundary.

Unfortunately it is not (yet) possible to completely excise this code from libssl still it is still required there in order to support legacy ENGINEs.

### CODE REVIEW

File: providers/implementations/ciphers/cipher_null.c [hunk 1]
```diff
@@ -87,6 +112,40 @@ static int null_get_ctx_params(void *vctx, OSSL_PARAM params[])
...
+    if (p != NULL
+        && !OSSL_PARAM_set_octet_ptr(p, ctx->tlsmac, ctx->tlsmacsize)) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+        return 0;
+    }
+    return 1;
+}
+
+static const OSSL_PARAM null_known_settable_ctx_params[] = {
+    OSSL_PARAM_END
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_null.c:125]: OSSL_CIPHER_PARAM_TLS_MAC_SIZE ?

File: ssl/record/tls_pad.c [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/rand.h>
+#include <openssl/evp.h>
+#include "internal/constant_time.h"
+#include "internal/cryptlib.h"
+
+/*
+ * This file has no dependencies on the reset of libssl because it is shared
```
Comment (slontis) [inline ssl/record/tls_pad.c:16]: rest

**Comment by slontis** [comment]
travis NIT is relevant.

File: ssl/record/tls_pad.c [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                                    unsigned char *recdata,
+                                    unsigned char **mac,
+                                    int *alloced,
+                                    size_t block_size, size_t mac_size,
+                                    int aead,
+                                    OPENSSL_CTX *libctx);
+
+/*-
+ * ssl3_cbc_remove_padding removes padding from the decrypted, SSLv3, CBC
+ * record in |rec| by updating |rec->length| in constant time. It also extracts
```
Comment (slontis) [inline ssl/record/tls_pad.c:53]: comment now needs updating..

File: ssl/record/tls_pad.c [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    good &= constant_time_ge_s(block_size, padding_length + 1);
+    *reclen -= good & (padding_length + 1);
+
+    return ssl3_cbc_copy_mac(reclen, origreclen, recdata, mac, alloced,
+                             block_size, mac_size, good, libctx);
+}
+
+/*-
+ * tls1_cbc_remove_padding removes the CBC padding from the decrypted, TLS, CBC
+ * record in |rec| in constant time. It also removes any explicit IV from the
```
Comment (slontis) [inline ssl/record/tls_pad.c:92]: comment is also incorrect here.

File: ssl/record/tls_pad.c [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                                    unsigned char *recdata,
+                                    unsigned char **mac,
+                                    int *alloced,
+                                    size_t block_size, size_t mac_size,
+                                    int aead,
+                                    OPENSSL_CTX *libctx)
+{
+    size_t good;
+    size_t padding_length, to_check, i;
+    size_t overhead = ((block_size == 1) ? 0 : 1) /* padding length byte */
```
Comment (slontis) [inline ssl/record/tls_pad.c:114]: Why is the ? 0 : 1 here now..

File: ssl/record/tls_pad.c [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+                      + mac_size;
+
+    /*
+     * These lengths are all public so we can test them in non-constant
+     * time.
+     */
+    if (overhead > *reclen)
+        return 0;
+
+    if (block_size != 1) {
```
Comment (slontis) [inline ssl/record/tls_pad.c:124]: what happened to the if (use_explicit_iv) code?

File: ssl/record/tls_pad.c [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+        good = constant_time_eq_s(0xff, good & 0xff);
+        *reclen -= good & (padding_length + 1);
+    }
+
+    return ssl3_cbc_copy_mac(reclen, origreclen, recdata, mac, alloced,
+                             block_size, mac_size, good, libctx);
+}
+
+/*-
+ * ssl3_cbc_copy_mac copies |md_size| bytes from the end of |rec| to |out| in
```
Comment (slontis) [inline ssl/record/tls_pad.c:173]: Same with this comment.

File: ssl/record/tls_pad.c [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+#define CBC_MAC_ROTATE_IN_PLACE
+
+static int ssl3_cbc_copy_mac(size_t *reclen,
+                             size_t origreclen,
+                             unsigned char *recdata,
+                             unsigned char **mac,
+                             int *alloced,
+                             size_t block_size,
+                             size_t mac_size,
+                             size_t good,
```
Comment (slontis) [inline ssl/record/tls_pad.c:196]: You dont seem to do anything with good..
It was used as the old return result for the caller (after some constant time mangling)

File: ssl/record/ssl3_record.c [hunk 9]
```diff
@@ -1175,67 +1178,129 @@ int tls1_enc(SSL *s, SSL3_RECORD *recs, size_t n_recs, int sending)
...
+                if (EVP_CIPHER_mode(enc) == EVP_CIPH_GCM_MODE) {
+                        recs[0].data += EVP_GCM_TLS_EXPLICIT_IV_LEN;
+                        recs[0].input += EVP_GCM_TLS_EXPLICIT_IV_LEN;
+                } else if (EVP_CIPHER_mode(enc) == EVP_CIPH_CCM_MODE) {
+                        recs[0].data += EVP_CCM_TLS_EXPLICIT_IV_LEN;
+                        recs[0].input += EVP_CCM_TLS_EXPLICIT_IV_LEN;
+                } else if (bs != 1 && SSL_USE_EXPLICIT_IV(s)) {
+                    recs[0].data += bs;
+                    recs[0].input += bs;
+                    recs[0].orig_len -= bs;
```
Comment (slontis) [inline ssl/record/ssl3_record.c:1215]: why not recs[0].length -= bs;

File: providers/implementations/ciphers/cipher_null.c (hunk 1 above)
Comment (mattcaswell) [reply providers/implementations/ciphers/cipher_null.c:126]: Fixed

File: ssl/record/tls_pad.c (hunk 2 above)
Comment (mattcaswell) [reply ssl/record/tls_pad.c]: Fixed

File: ssl/record/tls_pad.c (hunk 3 above)
Comment (mattcaswell) [reply ssl/record/tls_pad.c]: Fixed this and the other comments.

File: ssl/record/tls_pad.c (hunk 5 above)
Comment (mattcaswell) [reply ssl/record/tls_pad.c:124]: Some ciphers are actually have a block size of 1 and have no padding at all (and therefore no padding length byte) - but still have a MAC which we need to remove.

File: ssl/record/ssl3_record.c (hunk 9 above)
Comment (mattcaswell) [reply ssl/record/ssl3_record.c:1215]: Its already been updated immediately after the EVP_CipherUpdate call above

**Comment by mattcaswell** [comment]
All comments above addressed - as well as (hopefully) the travis nit. Please take another look.

**Comment by slontis** [comment]
travis error is relevant.

File: providers/implementations/ciphers/ciphercommon.c [hunk 10]
```diff
@@ -314,6 +411,15 @@ int cipher_generic_stream_update(void *vctx, unsigned char *out, size_t *outl,
     }
 
     *outl = inl;
+    /* Remove any TLS padding */
+    if (!ctx->enc && ctx->removetlspad > 0) {
```
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon.c:415]: Is this only supposed to be used by cipher_aes_cbc_hmac_sha1_hw.c & cipher_aes_cbc_hmac_sha256_hw.c.
Perhaps a comment would be good if this is the case.

File: providers/implementations/ciphers/cipher_null.c [hunk 11]
```diff
@@ -38,6 +53,14 @@ static OSSL_FUNC_cipher_cipher_fn null_cipher;
 static int null_cipher(void *vctx, unsigned char *out, size_t *outl,
                        size_t outsize, const unsigned char *in, size_t inl)
 {
+    PROV_CIPHER_NULL_CTX *ctx = (PROV_CIPHER_NULL_CTX *)vctx;
+
+    if (!ctx->enc && ctx->tlsmacsize > 0) {
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_null.c:58]: It probably wouldnt hurt to put in a ref comment.. e.g.

See https://tools.ietf.org/html/rfc5246#section-6.2.3.1

File: providers/implementations/ciphers/cipher_null.c [hunk 12]
```diff
@@ -38,6 +53,14 @@ static OSSL_FUNC_cipher_cipher_fn null_cipher;
...
+        if (inl < ctx->tlsmacsize)
+            return 0;
+        ctx->tlsmac = in + inl - ctx->tlsmacsize;
```
Comment (slontis) [inline providers/implementations/ciphers/cipher_null.c:61]: Is this right??

File: providers/implementations/ciphers/ciphercommon.c [hunk 13]
```diff
@@ -169,6 +183,8 @@ int cipher_generic_dinit(void *vctx, const unsigned char *key, size_t keylen,
                                         iv, ivlen, 0);
 }
 
+#define MAX_PADDING 256
```
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon.c:186]: If this includes the length then maybe comment on it..

File: providers/implementations/ciphers/ciphercommon.c [hunk 14]
```diff
@@ -178,6 +194,81 @@ int cipher_generic_block_update(void *vctx, unsigned char *out, size_t *outl,
...
+            if (outsize < inl + padnum) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+                return 0;
+            }
+
+            if (padnum > MAX_PADDING) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_CIPHER_OPERATION_FAILED);
+                return 0;
+            }
+            padval = (unsigned char)(padnum - 1);
```
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon.c:229]: what happens if padnum is 0?

File: providers/implementations/ciphers/cipher_null.c (hunk 11 above)
Comment (mattcaswell) [reply providers/implementations/ciphers/cipher_null.c:58]: Done

File: providers/implementations/ciphers/cipher_null.c (hunk 12 above)
Comment (mattcaswell) [reply providers/implementations/ciphers/cipher_null.c:65]: Yes, inl includes the MAC

File: providers/implementations/ciphers/ciphercommon.c (hunk 13 above)
Comment (mattcaswell) [reply providers/implementations/ciphers/ciphercommon.c:187]: Commment added

File: providers/implementations/ciphers/ciphercommon.c (hunk 10 above)
Comment (mattcaswell) [reply providers/implementations/ciphers/ciphercommon.c:419]: Comment added

**Comment by mattcaswell** [comment]
Fixups pushed to address latest feedback, and to resolve the Travis failure. Please take another look.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**